File: backend/python/text_extraction_service.py
#!/usr/bin/env python3
"""
Text Extraction Service - Extracts text from PDF, EPUB, and MOBI files

Usage:
    python text_extraction_service.py <file_path> <file_type>

Args:
    file_path: Path to the document file
    file_type: File extension (.pdf, .epub, .mobi)

Output:
    JSON object with extracted text:
    {"text": "Extracted text content..."}

Requirements:
    pip install pdfplumber ebooklib beautifulsoup4
"""

import sys
import json
import os

def extract_pdf(file_path):
    """Extract text from PDF using pdfplumber"""
    try:
        import pdfplumber
    except ImportError:
        return {"error": "pdfplumber not installed. Run: pip install pdfplumber"}

    text_parts = []

    try:
        with pdfplumber.open(file_path) as pdf:
            for page in pdf.pages:
                page_text = page.extract_text()
                if page_text:
                    text_parts.append(page_text)

        return {"text": "\n\n".join(text_parts)}
    except Exception as e:
        return {"error": f"PDF extraction failed: {str(e)}"}


def extract_epub(file_path):
    """Extract text from EPUB using ebooklib"""
    try:
        import ebooklib
        from ebooklib import epub
        from bs4 import BeautifulSoup
    except ImportError:
        return {"error": "ebooklib or beautifulsoup4 not installed. Run: pip install ebooklib beautifulsoup4"}

    text_parts = []

    try:
        book = epub.read_epub(file_path)

        for item in book.get_items():
            if item.get_type() == ebooklib.ITEM_DOCUMENT:
                # Parse HTML content
                soup = BeautifulSoup(item.get_content(), 'html.parser')
                text = soup.get_text(separator='\n', strip=True)
                if text:
                    text_parts.append(text)

        return {"text": "\n\n".join(text_parts)}
    except Exception as e:
        return {"error": f"EPUB extraction failed: {str(e)}"}


def extract_mobi(file_path):
    """Extract text from MOBI

    Note: MOBI is more complex. For now, return an error suggesting conversion.
    In the future, we could use mobi or KindleUnpack library.
    """
    return {"error": "MOBI extraction not yet implemented. Please convert to EPUB or PDF first."}


def main():
    if len(sys.argv) < 3:
        print(json.dumps({"error": "Usage: text_extraction_service.py <file_path> <file_type>"}))
        sys.exit(1)

    file_path = sys.argv[1]
    file_type = sys.argv[2].lower()

    # Check if file exists
    if not os.path.exists(file_path):
        print(json.dumps({"error": f"File not found: {file_path}"}))
        sys.exit(1)

    # Extract based on file type
    if file_type == '.pdf':
        result = extract_pdf(file_path)
    elif file_type == '.epub':
        result = extract_epub(file_path)
    elif file_type == '.mobi':
        result = extract_mobi(file_path)
    else:
        result = {"error": f"Unsupported file type: {file_type}"}

    # Output result as JSON
    print(json.dumps(result))

    # Exit with error code if extraction failed
    if "error" in result:
        sys.exit(1)
    else:
        sys.exit(0)


if __name__ == "__main__":
    main()

File: backend/src/analysis/text-extraction.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { DatabaseService } from '../database/database.service';
import * as fs from 'fs';
import * as path from 'path';
import { spawn } from 'child_process';

/**
 * TextExtractionService - Extracts text from documents (PDF, EPUB, TXT, etc.)
 *
 * This service handles extracting readable text from document files for:
 * - Searchability within the application
 * - Display in the info dialog
 * - Future AI analysis capabilities
 *
 * Supported formats:
 * - PDF: Uses pdfplumber (Python)
 * - EPUB: Uses ebooklib (Python)
 * - TXT/MD: Direct file read
 * - HTML/MHTML: Basic HTML text extraction
 */
@Injectable()
export class TextExtractionService {
  private readonly logger = new Logger(TextExtractionService.name);
  private readonly pythonScriptPath: string;

  constructor(private readonly databaseService: DatabaseService) {
    // Path to Python text extraction script
    this.pythonScriptPath = path.join(
      __dirname,
      '..',
      '..',
      'python',
      'text_extraction_service.py',
    );
  }

  /**
   * Extract text from a document file
   * @param mediaId - Media item ID
   * @param filePath - Path to the document file
   * @returns Extracted text content
   */
  async extractText(mediaId: string, filePath: string): Promise<string> {
    const ext = path.extname(filePath).toLowerCase();

    this.logger.log(`Extracting text from ${ext} file: ${filePath}`);

    try {
      let extractedText: string;

      if (ext === '.txt' || ext === '.md') {
        // Simple text files - just read directly
        extractedText = await this.extractPlainText(filePath);
      } else if (ext === '.pdf' || ext === '.epub' || ext === '.mobi') {
        // Use Python script for complex formats
        extractedText = await this.extractUsingPython(filePath, ext);
      } else if (ext === '.html' || ext === '.htm' || ext === '.mhtml') {
        // Basic HTML text extraction
        extractedText = await this.extractHtmlText(filePath);
      } else {
        throw new Error(`Unsupported file format: ${ext}`);
      }

      // Store extracted text in database
      this.databaseService.insertTextContent({
        mediaId,
        extractedText,
        extractionMethod: ext.replace('.', ''),
      });

      this.logger.log(
        `Text extraction complete: ${extractedText.length} characters`,
      );

      return extractedText;
    } catch (error: any) {
      this.logger.error(`Text extraction failed: ${error.message}`);
      throw error;
    }
  }

  /**
   * Extract text from plain text files (TXT, MD)
   */
  private async extractPlainText(filePath: string): Promise<string> {
    return fs.promises.readFile(filePath, 'utf-8');
  }

  /**
   * Extract text from HTML files
   * Strips HTML tags and returns plain text
   */
  private async extractHtmlText(filePath: string): Promise<string> {
    const htmlContent = await fs.promises.readFile(filePath, 'utf-8');

    // Basic HTML tag stripping (not perfect, but good enough for now)
    let text = htmlContent
      .replace(/<script[^>]*>[\s\S]*?<\/script>/gi, '') // Remove scripts
      .replace(/<style[^>]*>[\s\S]*?<\/style>/gi, '') // Remove styles
      .replace(/<[^>]+>/g, ' ') // Remove all HTML tags
      .replace(/&nbsp;/g, ' ') // Replace nbsp entities
      .replace(/&amp;/g, '&') // Replace amp entities
      .replace(/&lt;/g, '<') // Replace lt entities
      .replace(/&gt;/g, '>') // Replace gt entities
      .replace(/&quot;/g, '"') // Replace quot entities
      .replace(/&#039;/g, "'") // Replace apos entities
      .replace(/\s+/g, ' ') // Normalize whitespace
      .trim();

    return text;
  }

  /**
   * Extract text using Python script (for PDF, EPUB, MOBI)
   * NOTE: Requires additional Python packages to be installed:
   * - pip install pdfplumber ebooklib beautifulsoup4
   */
  private async extractUsingPython(
    filePath: string,
    fileType: string,
  ): Promise<string> {
    return new Promise((resolve, reject) => {
      // Check if Python script exists
      if (!fs.existsSync(this.pythonScriptPath)) {
        reject(
          new Error(
            `Python text extraction script not found at: ${this.pythonScriptPath}`,
          ),
        );
        return;
      }

      const pythonProcess = spawn('python3', [
        this.pythonScriptPath,
        filePath,
        fileType,
      ]);

      let stdout = '';
      let stderr = '';

      pythonProcess.stdout.on('data', (data) => {
        stdout += data.toString();
      });

      pythonProcess.stderr.on('data', (data) => {
        stderr += data.toString();
      });

      pythonProcess.on('close', (code) => {
        if (code !== 0) {
          this.logger.error(`Python extraction failed: ${stderr}`);
          reject(
            new Error(
              `Text extraction failed with code ${code}: ${stderr}`,
            ),
          );
          return;
        }

        try {
          // Python script returns JSON: { "text": "extracted content..." }
          const result = JSON.parse(stdout);
          resolve(result.text || '');
        } catch (error: any) {
          reject(
            new Error(`Failed to parse Python output: ${error.message}`),
          );
        }
      });

      pythonProcess.on('error', (error) => {
        reject(new Error(`Failed to spawn Python process: ${error.message}`));
      });
    });
  }

  /**
   * Get extracted text for a media item
   */
  async getExtractedText(mediaId: string): Promise<string | null> {
    const textContent = this.databaseService.getTextContent(mediaId);
    return textContent ? (textContent.extracted_text as string) : null;
  }

  /**
   * Check if text has been extracted for a media item
   */
  hasExtractedText(mediaId: string): boolean {
    const textContent = this.databaseService.getTextContent(mediaId);
    return textContent !== null;
  }

  /**
   * Re-extract text for a media item (useful if extraction failed previously)
   */
  async reExtractText(mediaId: string, filePath: string): Promise<string> {
    // Delete existing text content
    this.databaseService.deleteTextContent(mediaId);

    // Extract again
    return this.extractText(mediaId, filePath);
  }
}
